dot/network/light.go

```diff
@@ -6,8 +6,77 @@ import (
 	"github.com/ChainSafe/gossamer/dot/types"
 	"github.com/ChainSafe/gossamer/lib/common"
 	"github.com/ChainSafe/gossamer/pkg/scale"
+
+	libp2pnetwork "github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
 )
 
+// handleLightStream handles streams with the <protocol-id>/light/2 protocol ID
+func (s *Service) handleLightStream(stream libp2pnetwork.Stream) {
+	s.readStream(stream, s.decodeLightMessage, s.handleLightMsg)
+}
+
+func (s *Service) decodeLightMessage(in []byte, peer peer.ID, _ bool) (Message, error) {
+	s.lightRequestMu.RLock()
+	defer s.lightRequestMu.RUnlock()
+
+	// check if we are the requester
+	if _, requested := s.lightRequest[peer]; requested {
+		// if we are, decode the bytes as a LightResponse
+		msg := NewLightResponse()
+		err := msg.Decode(in)
+		return msg, err
+	}
+
+	// otherwise, decode bytes as LightRequest
+	msg := NewLightRequest()
+	err := msg.Decode(in)
+	return msg, err
+}
+
+func (s *Service) handleLightMsg(stream libp2pnetwork.Stream, msg Message) error {
+	defer func() {
+		_ = stream.Close()
+	}()
+
+	lr, ok := msg.(*LightRequest)
+	if !ok {
+		return nil
```

Comment:
it should *never* not be a request, just by reasoning of how the substrate networking protocol works (https://docs.rs/sc-network/0.9.0/sc_network/) if it's not a request, then we can ignore it, as it shouldn't be sent over an inbound stream. (in fact we can probably downscore the peer, but there's another issue open for that)

Reply:
we can't change it to be `*LightRequest` as it needs to implement the function type `messageHandler = func(stream libp2pnetwork.Stream, msg Message) error`

Reply:
Thanks for clarifying!!

Reply:
Actually I think we return an error since it's not meant to happen.

Or just remove the type assertion check so it panics, up to you. Panicing is fine too in this case I'd say.